src/elliptic/curves/wrappers.rs

```diff
@@ -0,0 +1,1692 @@
+    pub fn y_coord(&self) -> BigInt {
+        self.as_point_ref().y_coord()
+    }
+
+    /// Adds two points, returns the result, or `None` if resulting point is zero
+    pub fn add_checked(&self, point: PointRef<E>) -> Result<Self, ZeroPointError> {
+        self.as_point_ref().add_checked(point)
+    }
+
+    /// Substrates two points, returns the result, or `None` if resulting point is zero
+    pub fn sub_checked(&self, point: PointRef<E>) -> Result<Self, ZeroPointError> {
+        self.as_point_ref().sub_checked(point)
+    }
+
+    /// Multiplies a point at scalar, returns the result, or `None` if resulting point is zero
+    pub fn mul_checked_z(&self, scalar: &ScalarZ<E>) -> Result<Self, ZeroPointError> {
+        self.as_point_ref().mul_checked_z(scalar)
+    }
+
+    /// Multiplies a point at nonzero scalar, returns the result, or `None` if resulting point is zero
+    pub fn mul_checked(&self, scalar: &Scalar<E>) -> Result<Self, ZeroPointError> {
+        self.as_point_ref().mul_checked(scalar)
+    }
+
+    /// Serializes point into (un)compressed form
+    pub fn to_bytes(&self, compressed: bool) -> Vec<u8> {
+        self.as_point_ref().to_bytes(compressed)
+    }
+
+    /// Creates [PointRef] that holds a reference on `self`
+    pub fn as_point_ref(&self) -> PointRef<E> {
+        PointRef::from(self)
+    }
+
+    fn from_raw(raw_point: E::Point) -> Result<Self, ZeroPointError> {
+        if raw_point.is_zero() {
+            Err(ZeroPointError(()))
+        } else {
+            Ok(Self { raw_point })
+        }
+    }
+
+    unsafe fn from_raw_unchecked(point: E::Point) -> Self {
+        Point { raw_point: point }
+    }
+
+    fn as_raw(&self) -> &E::Point {
+        &self.raw_point
+    }
+
+    fn into_raw(self) -> E::Point {
+        self.raw_point
+    }
+}
+
+impl<E: Curve> Clone for Point<E> {
+    fn clone(&self) -> Self {
+        //  Safety: `self` is guaranteed to be non-zero
+        unsafe { Point::from_raw_unchecked(self.as_raw().clone()) }
+    }
+}
+
+impl<E: Curve> fmt::Debug for Point<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.as_raw().fmt(f)
+    }
+}
+
+impl<E: Curve> TryFrom<PointZ<E>> for Point<E> {
+    type Error = ZeroPointError;
+    fn try_from(point: PointZ<E>) -> Result<Self, Self::Error> {
+        Self::from_raw(point.into_raw())
+    }
+}
+
+/// Elliptic curve generator
+///
+/// Holds internally a static reference on curve generator. Can be used in arithmetic interchangeably
+/// as [`PointRef<E>`](PointRef).
+///
+/// Generator multiplied at non-zero scalar always produce non-zero point, thus output type of
+/// the multiplication is [`Point<E>`](Point). This is the only difference compared to `Point<E>` and
+/// `PointRef<E>`. Use [`to_point_owned`](Self::to_point_owned) and [`as_point_ref`](Self::as_point_ref)
+/// methods to convert the generator into `Point<E>` and `PointRef<E>`.
+pub struct Generator<E: Curve> {
```

Comment:
I think this should be something that the curves can specialize, because most curve libs have constant time high performance generator multiplication,
so IMHO we should either make a generator trait, or add functions to the Curve trait for generator operations

Reply:
Thank for pointing me out! I will add methods to the ECPoint trait (I think it fits more naturally). Should I provide methods just for generator multiplication, or curve implementation might want to overload generator addition/subtraction as well?

Reply:
I think just multiplying by the generator is enough

Reply:
Generator multiplication is optimised: https://github.com/ZenGo-X/curv/pull/120/commits/ce61396c7583e847a17b43ea1731306d280f5df8

Reply:
Nice :) (I have some other optimizations I can see here but I think it's better to open a PR after this and not make this any more complicated)